### PR TITLE
WIP: PCA Tool Checkin

### PR DESCRIPTION
**Pull request by lorrainealisha75**

### CODE REVIEW

File: tools/sklearn/pca.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            pca_input = np.dot(pca_input, pca_input.T)
+
+        if args.eigen_solver == 'arpack':
+            pca_params.update({'tol': args.tolerance, 'max_iter': args.max_iter})
+
+        pca = KernelPCA()
+
+    pca.set_params(**pca_params)
+    pca_output = pca.fit_transform(pca_input)
+    np.savetxt(fname=args.outfile, X=pca_output, fmt='%.4f')
```
Comment (simonbray) [inline tools/sklearn/pca.py:92]:
```suggestion
    np.savetxt(fname=args.outfile, X=pca_output, delimiter='\t')
```
@lorrainealisha75, can you use a tab as the delimiter? Galaxy expects it for tabular files and hopefully then the has_n_columns test will pass.

**Comment by simonbray** [comment]
I wrote a couple of comments yesterday which disappeared for some reason, here they are again:
* the input section could use the sklearn main_macros.xml to allow selection of columns by header name as well as index
* the input file should be formatted with tab as the separator

File: tools/sklearn/pca.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                        help="Base name for output file (no extension).")
+    args = parser.parse_args()
+
+    usecols = None
+    cols = []
+    pca_params = {}
+
+    if args.columns != 'all':
+        with open(args.infile) as f:
+            num_cols = len(f.readline().split(' '))
```
Comment (simonbray) [inline tools/sklearn/pca.py:49]:
```suggestion
            num_cols = len(f.readline().split())
```

**Comment by simonbray** [comment]
@khanteymoori, can you have a look? There are a couple of tests which aren't passing yet but I think it should be ready as far as the user interface is concerned.

Thanks a lot @lorrainealisha75!

**Comment by simonbray** [comment]
Tests are passing: https://travis-ci.org/github/bgruening/galaxytools/builds/694642169

File: tools/sklearn/pca.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            pca_params.update({'coef0': args.coef0})
+        elif args.kernel == 'precomputed':
+            pca_input = np.dot(pca_input, pca_input.T)
+
+        if args.eigen_solver == 'arpack':
+            pca_params.update({'tol': args.tolerance, 'max_iter': args.max_iter})
+
+        pca = KernelPCA()
+
+    print(pca_params)
```
Comment (bgruening) [inline tools/sklearn/pca.py:92]: is this needed?

File: tools/sklearn/test-data/pca_kernel_output2.dat [hunk 4]
```diff
@@ -0,0 +1,300 @@
+-789.0683 244.4933 86.3224 14.2457 336.2516 -48.3858 104.9891 115.2530
+-721.5195 53.8094 164.5249 -213.6399 283.9916 -180.7496 -85.6662 -56.4151
+-731.7468 192.1683 -263.2790 -109.8660 -131.7332 -389.4955 60.0203 53.8620
+-789.8859 246.9337 102.2221 -46.5027 312.8768 -90.5573 -21.4048 101.4942
+-818.3874 350.2119 -440.5530 165.6495 224.8707 93.6159 -25.2826 7.4475
+-833.1559 459.6316 -1178.7885 276.9423 -47.5942 118.5477 -287.9134 -43.3138
```
Comment (bgruening) [inline tools/sklearn/test-data/pca_kernel_output2.dat:6]: is this really a tabular file?

File: tools/sklearn/pca.py (hunk 3 above)
Comment (simonbray) [reply tools/sklearn/pca.py:92]: no, but it's handy to have in the stdout IMO

File: tools/sklearn/test-data/pca_kernel_output2.dat (hunk 4 above)
Comment (simonbray) [reply tools/sklearn/test-data/pca_kernel_output2.dat:6]: @lorrainealisha75, could you also look at this? The output files should be tab-separated.

**Comment by simonbray** [comment]
Is this okay to merge @khanteymoori @bgruening (when the tab-separated thing is fixed)? Obviously there are a huge number of things that could be added but at least this provides a basic functionality.

**Comment by bgruening** [comment]
Looks good from my side if the tab thing is fixed.

**Comment by khanteymoori** [comment]
@simonbray let me test it again over the weekend with more data to make sure it produces reliable results. 
thanks

**Comment by simonbray** [comment]
Would be great, thanks a lot.

**Pull request closed without merging**